### PR TITLE
Java 8 and kotlin

### PR DESCRIPTION
**Pull request by rpmoore**
This pull request updates the following:

* Updates the project to java 8
* Adds Kotlin 1.1.4
* Updated gradle to 4.1
* Updates may dependencies to their latest versions
* Consolidated most of the version information into the `gradle.properties` file
* Code cleanup removing unused imports and removing redundant type declarations
* Adds a paginated bucket delete command
* Added some missing finals
* Re-implements the file walking helper method to be lazy
* Fixed some of the start bulk operation helper functions to make them truly lazy
* Updates to the shadowjar setup to correctly include/exclude dependencies

### CODE REVIEW

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/utils/collections/StreamWrapper.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.spectralogic.ds3client.utils.collections;
```
Comment (RachelTucker) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/utils/collections/StreamWrapper.java]: missing copyright

File: ds3-sdk/src/main/kotlin/com/spectralogic/ds3client/helpers/DeleteBucket.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.spectralogic.ds3client.helpers
```
Comment (RachelTucker) [inline ds3-sdk/src/main/kotlin/com/spectralogic/ds3client/helpers/DeleteBucket.kt]: missing copyright

File: ds3-sdk/src/main/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+object FileTreeWalker {
+    /**
+     * Walks a file system starting at {@param startDir} and returns a Lazy Iterable with all the
+     * files represented as Ds3Objects.
+     */
+    fun walk(startDir: Path) : Iterable<Ds3Object> {
+        return StreamWrapper.wrapStream(Files.walk(startDir).map {
+            Ds3Object(startDir.relativize(it).toString().replace("\\", "/"), Files.size(it))
+        }.filter {
+            !Guard.isStringNullOrEmpty(it.name)  // skip entries that are null or are the empty string
```
Comment (GraciesPadre) [inline ds3-sdk/src/main/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker.kt]: Since this is kotlin, why not just  ! it.name.isNullOrEmpty()?

Comment (rpmoore) [reply ds3-sdk/src/main/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker.kt]: I didn't realize kotlin had that.  Updating...

File: ds3-sdk/src/test/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker_Test.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        tempDir.newFile("bar")
+        tempDir.newFile("baz")
+
+        val walk = FileTreeWalker.walk(tempDir.root.toPath())
+
+        val listBuilder = ImmutableList.Builder<Ds3Object>()
+
+        walk.forEach {
+            listBuilder.add(it)
+            println(it.name)
```
Comment (GraciesPadre) [inline ds3-sdk/src/test/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker_Test.kt]: Log, as opposed to println()?

Comment (rpmoore) [reply ds3-sdk/src/test/kotlin/com/spectralogic/ds3client/helpers/FileTreeWalker_Test.kt]: Opps. That was a debug print.  I'll remove

**Pull request merged**